src/NodaTime/Utility/Cache.cs

```diff
@@ -40,68 +43,46 @@ internal Cache(int size, Func<TKey, TValue> valueFactory, IEqualityComparer<TKey
         /// <returns>The value associated with the key.</returns>
         internal TValue GetOrAdd(TKey key)
         {
-            lock (mutex)
+            if (dictionary.TryGetValue(key, out var value))
             {
-                // First check the cache...
-                if (dictionary.TryGetValue(key, out TValue value))
-                {
-                    return value;
-                }
-
-                // Make space if necessary...
-                if (dictionary.Count == size)
-                {
-                    TKey firstKey = keyList.First.Value;
-                    keyList.RemoveFirst();
-                    dictionary.Remove(firstKey);
-                }
-
-                // Create and cache the new value
-                value = valueFactory(key);
-                keyList.AddLast(key);
-                dictionary[key] = value;
                 return value;
             }
+            // Add the key to the eviction queue is we're *probably* going to be adding it now.
+            // There's no easy way of telling whether a particular call to GetOrAdd has added the key.
+            keyList.Enqueue(key);
+            value = dictionary.GetOrAdd(key, valueFactory);
+
+            // Trim to size if necessary, bearing in mind that other threads may be doing
+            // the same thing at the same time, and that there may be multiple entries for a particular key.
+            while (dictionary.Count > size && keyList.TryDequeue(out var keyToRemove))
+            {
+                dictionary.TryRemove(keyToRemove, out _);
+            }
+            return value;
         }
 
         /// <summary>
         /// Returns the number of entries currently in the cache, primarily for diagnostic purposes.
         /// </summary>
-        internal int Count
-        { 
-            get 
-            {
-                lock (mutex)
-                {
-                    return dictionary.Count;
-                }
-            }
-        }
+        internal int Count => dictionary.Count;
 
         /// <summary>
         /// Returns a copy of the keys in the cache as a list, for diagnostic purposes.
         /// </summary>
-        internal List<TKey> Keys
-        {
-            get
-            {
-                lock (mutex)
-                {
-                    return new List<TKey>(keyList);
-                }
-            }
-        }
+        internal List<TKey> Keys => dictionary.ToArray().Select(pair => pair.Key).ToList();
```

Comment:
`ConcurrentDictionary<,>.Keys` is very poorly documented - I like the fact that by calling `ToArray` it's obvious that we're getting a snapshot, and the dictionary gets to do that immediately and "in the open" rather than potentially creating a snapshot behind the scenes. If this were non-diagnostic code, I'd benchmark... but in this case it doesn't really matter much :)